app/models/smart_proxy.rb

```diff
@@ -82,18 +82,34 @@ def refresh
   end
 
   def taxonomy_foreign_conditions
-    feature_names = self.features.pluck(:name)
-    conditions                      = {}
-    if feature_names.include?('Puppet') && feature_names.include?('Puppet CA')
+    conditions = {}
+    if has_puppet? && has_puppet_ca?
       conditions = "puppet_proxy_id = #{id} OR puppet_ca_proxy_id = #{id}"
-    elsif feature_names.include?('Puppet')
+    elsif has_puppet?
       conditions[:puppet_proxy_id] = id
-    elsif feature_names.include?('Puppet CA')
+    elsif has_puppet_ca?
       conditions[:puppet_ca_proxy_id] = id
     end
     conditions
   end
 
+  # Helper methods for features - e.g. has_puppet? has_dns?
+  def method_missing(method, *args, &block)
```

Comment:
I find magic like this less reliable than an explicit method, which is easy to understand and test.  It also avoids needing to update method prototypes for core Ruby methods if they change (`#respond_to?` has before).

Rails tends to use define_method as far as I know, but that then introduces the class-load issue I mentioned in the ticket.